### PR TITLE
Im gonna dooocument

### PR DESCRIPTION
**Pull request by Quaqqer**
- Some documentation for data types used in the Ast and errors
  - Some of the errors types I didn't really understand and found very few usages of, so I didn't understand their purpose
- Pretty print records

### CODE REVIEW

File: src/ast.rs [hunk 1]
```diff
@@ -1,38 +1,75 @@
...
 pub struct ProperName<'t>(pub &'t str, pub Span);
 
+/// An outer definition
 #[derive(Debug, Clone)]
 pub enum Def<'t> {
+  /// A normal definition
+  ///
+  /// Example
+  /// ```sylt
+  /// a : Int := 1
```
Comment (FredTheDino) [inline src/ast.rs]:
```suggestion
  /// def a : Int := 1
```

File: src/ast.rs [hunk 2]
```diff
@@ -84,13 +146,20 @@ impl<'t> Pattern<'t> {
...
 
+/// One constructor of an enum
 #[derive(Debug, Clone)]
 pub struct EnumConst<'t> {
   pub tag: ProperName<'t>,
   pub ty: Option<Type<'t>>,
   pub span: Span,
 }
 
+/// A "which" branch in a match expression
```
Comment (FredTheDino) [inline src/ast.rs]:
```suggestion
/// A "with" branch in a match expression
```

File: src/ast.rs [hunk 3]
```diff
@@ -99,33 +168,70 @@ pub struct WithBranch<'t> {
...
+  /// An integer value
   EInt(i64, Span),
+
+  /// A real number value
   EReal(f64, Span),
+
+  /// A string value
   EStr(&'t str, Span),
 
+  /// An enum value
```
Comment (FredTheDino) [inline src/ast.rs]:
```suggestion
  /// An enum constructor value
```

File: src/ast.rs [hunk 4]
```diff
@@ -189,19 +305,32 @@ impl UnOp {
...
   And(Span),
+  /// Or operator `or`
   Or(Span),
+  /// Less than operator `<`
   Lt(Span),
+  /// Less than or equal operator `<=`
   LtEq(Span),
+  /// Equality operator `==`
   Eq(Span),
+  /// Inequality operator `~=`
```
Comment (FredTheDino) [inline src/ast.rs]:
```suggestion
  /// Inequality operator `!=`
```

File: src/error.rs [hunk 5]
```diff
@@ -191,7 +193,7 @@ impl Error {
       ),
 
       Error::ResUnknown { name, span } => format!(
-        "> I couldn't figure out what {:?} refernces. Did you make a typo?\n{}",
+        "> I couldn't figure out what {:?} references. Did you make a typo?\n{}",
```
Comment (FredTheDino) [inline src/error.rs:196]: <3

File: src/error.rs [hunk 6]
```diff
@@ -17,83 +17,83 @@ impl Display for Span {
...
+  ResNoEnum { ty_name: String, at: Span },
 
-  ResMsg {
-    msg: String,
-    span: Span,
-  },
+  /// FIXME: ?
+  ResMsg { msg: String, span: Span },
 
+  /// FIXME: ?
```
Comment (FredTheDino) [inline src/error.rs:53]: All `Check*` messages are type-errors. So this is just a generic type error.

File: src/error.rs [hunk 7]
```diff
@@ -17,83 +17,83 @@ impl Display for Span {
...
   CheckMsg {
     msg: &'static str,
     a_span: Span,
     b_span: Span,
   },
 
+  /// FIXME: ?
```
Comment (FredTheDino) [inline src/error.rs:60]: When we expect a type but get something else.

File: src/preamble.lua [hunk 8]
```diff
@@ -5,11 +5,34 @@ function Enum.__tostring(t)
 end
 
 function Enum.new(k, v)
-  local e = {k, v}
+  local e = { k, v }
```
Comment (FredTheDino) [inline src/preamble.lua:14]: Agressive

File: src/preamble.lua [hunk 9]
```diff
@@ -5,11 +5,34 @@ function Enum.__tostring(t)
...
   setmetatable(e, Enum)
   return e
 end
 
+RecordMeta = {}
+function RecordMeta.__tostring(record)
```
Comment (FredTheDino) [inline src/preamble.lua]: I'd love to have tests for this, I can show you how to do that if you don't know how, but it's pretty straight forward. But you need a `tostring` foreign function and then just check the equality using pattern matches. Let me know if you need help. :D

**Review by FredTheDino — COMMENTED** [review]
Might fine! Just some minor nits.

File: src/preamble.lua (hunk 8 above)
Comment (Quaqqer) [reply src/preamble.lua:14]: I would blame it on my formatter, but that was probably me ;)

**Pull request merged**